tokio/tests/sync_rwlock.rs

```diff
@@ -0,0 +1,163 @@
+#![warn(rust_2018_idioms)]
+
+use std::sync::Arc;
+
+use futures_util::future::FutureExt;
+use futures_util::stream::StreamExt;
+use futures_util::try_future::TryFutureExt;
+use futures_util::{future, stream};
+
+use tokio::sync::{oneshot, RwLock};
+use tokio_test::task::spawn;
+use tokio_test::{assert_pending, assert_ready};
+
+// multiple reads should be Ready
+#[test]
+fn read_shared_ready() {
+    let rwlock = RwLock::new(100);
+    let mut t1 = spawn(rwlock.read());
+    assert_ready!(t1.poll());
+    let mut t2 = spawn(rwlock.read());
+    assert_ready!(t2.poll());
+}
+
+// when there is an active shared ownded, exclusive access should not be possible
+#[test]
+fn write_shared_pending() {
+    let rwlock = RwLock::new(100);
+    let mut t1 = spawn(rwlock.read());
+    let _g1 = assert_ready!(t1.poll());
+    let mut t2 = spawn(rwlock.write());
+    assert_pending!(t2.poll());
+}
+
+// Acquire an RwLock nonexclusively by a single task
+#[tokio::test]
+async fn read_uncontested() {
+    let rwlock = RwLock::new(100);
+    let result = *rwlock.read().await;
+
+    assert_eq!(result, 100);
+}
+
+// Acquire an RwLock nonexclusively by two different tasks simultaneously .
+#[tokio::test]
+async fn read_shared() {
+    let rwlock = RwLock::new(100);
+
+    let (tx0, rx0) = oneshot::channel::<()>();
+    let (tx1, rx1) = oneshot::channel::<()>();
+    let task0 = rwlock.read().then(move |guard| {
+        tx1.send(()).unwrap();
+        rx0.map(move |_| *guard)
+    });
+    let task1 = rwlock.read().then(move |guard| {
+        tx0.send(()).unwrap();
+        rx1.map(move |_| *guard)
+    });
+
+    let result = future::join(task0, task1).await;
```

Comment:
I don't think this actually tests that the two locks are held simultaneously?

Reply:
don't they have to? since both reads need to wait for the other channel message to release the guard

Reply:
I think it's better to write tests like this by polling the futures directly, rather than trying to use await.  I've got a version of this test that I haven't posted yet; I'll try to post it tonight.

Reply:
yeah totally makes sense, updated!

Reply:
turns out there was already a test for that, i am going to remove `read_shared` and rename `read_shared_ready` to `read_shared`